### PR TITLE
Update ListProof hashing rules [ECR-3318]

### PR DESCRIPTION
**Pull request by MakarovS**
## Overview
Update ListProof hashing rules.
Should we add `length` field to `ListProofBranch` as a way to pass length of the index?
WIP: length is always `1` when calculating hash right now

---
See: https://jira.bf.local/browse/ECR-3318

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has Javadoc
- [x] Method preconditions are checked and documented in the Javadoc of the method
- [x] Changelog is updated if needed (in case of notable or breaking changes)
- [x] The continuous integration build passes

### CODE REVIEW

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculator.java [hunk 1]
```diff
@@ -34,7 +36,14 @@
  *
  * @param <E> the type of elements in the corresponding list
  */
-final class ListProofRootHashCalculator<E> implements ListProofVisitor {
+final class ListMerkleRootCalculator<E> implements ListProofVisitor {
```
Comment (dmitry-timofeev) [inline exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculator.java]: Update the docs ^

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculator.java [hunk 2]
```diff
@@ -44,26 +53,33 @@
...
     elements = new TreeMap<>();
     index = 0;
-    calculatedRootHash = null;
+    merkleRoot = null;
+
+    ListProofNode rootNode = listProof.getRootNode();
+    rootNode.accept(this);
 
-    listProof.accept(this);
+    merkleRoot = hashFunction.newHasher()
```
Comment (dmitry-timofeev) [inline exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculator.java]: It turns out to be **not** the Merkle root, but the hash of the collection (what is the term for that?).

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculator.java [hunk 3]
```diff
@@ -120,14 +140,14 @@ private HashCode visitLeft(ListProofBranch branch, long parentIndex) {
    *
    * @return hash code
```
Comment (dmitry-timofeev) [inline exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculator.java]: The documentation is not complete + contains redundant @return

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListProof.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package com.exonum.binding.common.proofs.list;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Represents a proof that some elements exist in a ProofList at certain positions and the length
+ * of corresponding index.
+ */
+public class ListProof {
```
Comment (dmitry-timofeev) [inline exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListProof.java]: final

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/UncheckedListProof.java [hunk 5]
```diff
@@ -30,5 +30,5 @@
   /**
    * Returns raw source proof of this UncheckedListProof.
    */
-  ListProofNode getRootProofNode();
+  ListProof getListProof();
```
Comment (dmitry-timofeev) [inline exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/UncheckedListProof.java]: I don't quite get what is going on here. Why isn't ListProof UncheckedListProof,
  but is contained in it?

File: exonum-java-binding/common/src/test/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculatorTest.java [hunk 6]
```diff
@@ -56,14 +60,16 @@ void visit_FullProof2elements() {
     ListProofElement right = leafOf(V2);
     ListProofBranch root = new ListProofBranch(left, right);
 
-    calculator = createListProofCalculator(root);
+    ListProof listProof = new ListProof(root, 2);
+    calculator = createListProofCalculator(listProof);
 
-    //calculate expected root hash
+    // Calculate expected merkle root
```
Comment (dmitry-timofeev) [inline exonum-java-binding/common/src/test/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculatorTest.java]: Here and elsewhere, please use consistent terms: merkle root for the hash of the tree, and (? — don't remember) hash for the hash of the collection.

File: exonum-java-binding/core/rust/src/storage/proof_list_index.rs [hunk 7]
```diff
@@ -369,6 +369,19 @@ pub extern "system" fn Java_com_exonum_binding_core_storage_indices_ProofListInd
     handle::drop_handle::<ProofListIndexIter<Value>>(&env, iter_handle);
 }
 
+fn make_java_proof_root<'a>(
+    env: &JNIEnv<'a>,
+    proof: &ListProof<Value>,
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/src/storage/proof_list_index.rs]: @vitvakatu  Why doesn't ListProof in Rust contain the length?

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculator.java (hunk 2 above)
Comment (pavel-mukhanov) [reply exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListMerkleRootCalculator.java]: It's `hash` of the ProofList object in MerkleDB.

File: exonum-java-binding/core/rust/src/storage/proof_list_index.rs (hunk 7 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/src/storage/proof_list_index.rs]: I have no idea https://github.com/exonum/exonum/blob/master/components/merkledb/src/proof_list_index/proof.rs#L144

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/UncheckedListProof.java (hunk 5 above)
Comment (MakarovS) [reply exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/UncheckedListProof.java]: `ListProof` is an entity that contains `ListProofNode` and `length`. We could go with a different name for that entity, that's the best one I could come up with. I thought about `ListProofRoot` naming, but it doesn't reflect the fact that we have length in there.

Comment (dmitry-timofeev) [reply exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/UncheckedListProof.java]: ListProof name is perfect — the question is why we have a ListProof which seems to be a complete proof from the list — and also an UnchekedListProof interface that contains the ListProof.

File: exonum-java-binding/core/rust/src/storage/proof_list_index.rs (hunk 7 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/src/storage/proof_list_index.rs]: That is really strange — why force a user provide the size when we assume they don't know _anything_ about the list (including its size!), don't trust the system, hence need a proof. How do they know the size (without trusting the system)? If they do trust it, then the proof is not needed at all.

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListProofHashCalculator.java [hunk 8]
```diff
@@ -116,18 +136,16 @@ private HashCode visitLeft(ListProofBranch branch, long parentIndex) {
   }
 
   /**
-   * Returns calculated root hash of a list proof tree.
-   *
-   * @return hash code
+   * Returns calculated hash of a list proof tree.
```
Comment (dmitry-timofeev) [inline exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/ListProofHashCalculator.java]: The documentation is still incomplete: is it a merkle root hash of the tree, or a hash of the ProofList collection?

File: exonum-java-binding/core/rust/src/storage/proof_list_index.rs (hunk 7 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/src/storage/proof_list_index.rs]: Yep, looks like it violates our privacy guarantees. Created https://jira.bf.local/browse/ECR-3414

Comment (pavel-mukhanov) [reply exonum-java-binding/core/rust/src/storage/proof_list_index.rs]: Maybe I missing something, but in rust, we force the user to provide list length only when we validating the proof. We don't need list size to construct the proof.

Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/src/storage/proof_list_index.rs]: > we force the user to provide list length only when we validating the proof.

Why does the core exclude the essential part of the proof? The proof cannot be verified without this piece of data. Why is the user forced to obtain that missing piece of the proof manually?

File: exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/UncheckedListProof.java (hunk 5 above)
Comment (MakarovS) [reply exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/UncheckedListProof.java]: Created a task for that - https://jira.bf.local/browse/ECR-3429

**Comment by dmitry-timofeev** [comment]
It doesn't build

**Comment by MakarovS** [comment]
Fixed that.
Although now it gets 405 from coveralls -_- restarted the job, should work now.

**Comment by dmitry-timofeev** [comment]
Good, thanks, coveralls had been having some issues, failing all reporting jobs.

**Pull request merged**